docs/src/gradient_accumulation.md

```diff
@@ -0,0 +1,134 @@
+For this reason every inplaceable thunk has both a `val` field holding the value representation, and a `add!` field holding the action representation.
+The `val` field use a plain [`Thunk`](@ref) to avoid the computation (and thus allocation) for if it is unused.
+
+!!! note "Do we need both representations?"
+    Right now every [`InplaceableThunk`](@ref) has two fields that need to be specified.
+    The value form (represented as a the `Thunk`(@ref) typed field), and the action form (represented as the `add!` field).
+    It is possible in a future version of ChainRulesCore.jl we will work out a clever way to find the zero differential for arbitrary primal values.
+    Given that we could always just determine the value form from `inplaceable.add!(zero_differential(primal))`.
+    There are some technical difficulties in finding the zero differentials, but this may be solved at some point.
+
+
+The `+` operation on `InplaceableThunk`s is overloaded to [`unthunk`](@ref) that `val` field to get the value form.
+Where as the [`add!!`](@ref) operation is overloaded to call `add!` to invoke the action.
+
+With `getindex` defined to return an `InplaceableThunk`, we now get to `X̄ = add!!(ā, b̄)` requires only a single allocation.
+This allocation occurs when `unthunk`ing `ā`, which is then mutated to become `X̄`.
+This is basically as good as we can get: if we want `X̄` to be an `Array` then at some point we need to allocate that array.
+
+!!! note "Can we do more? Deferred accumulation"
+    We could keep going further to drop allocations if we really wanted.
+    If we didn't care about `X̄` being an `Array` then we could defer its computation too.
+    `X̄ = @thunk add!!(ā, b̄)`.
+    This kind of deferral will work fine and you can keep chaining it.
+    It does start to burn stack space, and might make the compiler's optimization passes cry.
+    But it's valid and should work fine.
+
+### Examples of InplaceableThunks
+
+#### `getindex`
+
+The aforementioned `getindex` is really the poster child for this.
+Consider something like:
+```julia
+function mysum(X::Array{Float64})
+    total = 0.0
+    for i in eachindex(X)
+        total += X[i]
+    end
+    return total
+end
+```
+If one only has value representation of derivatives one ends up having to allocate a derivative array for every single element of the original array `X`.
+That's terrible.
+Where-as with the action representation that `InplaceableThunk`s provide, there is just a single `Array` allocated.
+One can see [the `getindex` rule in ChainRules.jl for the implementation](https://github.com/JuliaDiff/ChainRules.jl/blob/v0.7.49/src/rulesets/Base/indexing.jl).
+
+
+#### matmul etc (`*`)
+Multiplication of scalars/vectors/matrixes of compatible dimentions can all also have their derivatives represented as a `InplaceableThunk`.
+These tend to pivot around that `add!` action being defined along the lines of:
```

Comment:
I struggle to understand the rest of the paragraph. What allocations are we avoiding here?

Reply:
we avoid allocating the matrix `A'*Ȳ`
because we directly perform `mul!(X̄, A', Ȳ, true, true)`
probably some edit here needed to make clear